python/seldon_core/batch_processor.py

```diff
@@ -0,0 +1,409 @@
+import click
+import json
+import requests
+from queue import Queue
+from threading import Thread
+from seldon_core.seldon_client import SeldonClient
+import numpy as np
+import os
+import uuid
+import time
+
+CHOICES_GATEWAY_TYPE = ["ambassador", "istio", "seldon"]
+CHOICES_TRANSPORT = ["rest", "grpc"]
+CHOICES_PAYLOAD_TYPE = ["ndarray", "tensor", "tftensor"]
+CHOICES_DATA_TYPE = ["data", "json", "str"]
+CHOICES_METHOD = ["predict"]
+CHOICES_LOG_LEVEL = ["debug", "info", "warning", "error"]
+
+
+def start_multithreaded_batch_worker(
+    deployment_name: str,
+    gateway_type: str,
+    namespace: str,
+    host: str,
+    transport: str,
+    data_type: str,
+    payload_type: str,
+    workers: int,
+    retries: int,
+    input_data_path: str,
+    output_data_path: str,
+    method: str,
+    log_level: str,
+    benchmark: bool,
+    batch_id: str,
+) -> None:
+    """
+    Starts the multithreaded batch worker which consists of three worker types and
+    two queues; the input_file_worker which reads a file and puts all lines in an
+    input queue, which are then read by the multiple request_processor_workers (the
+    number of parallel workers is specified by the workers param), which puts the output
+    in the output queue and then the output_file_worker which puts all the outputs in the
+    output file in a thread-safe approach.
+
+    All parameters are defined and explained in detail in the run_cli function.
+    """
+    start_time = time.time()
+
+    q_in = Queue(workers * 2)
+    q_out = Queue(workers * 2)
+
+    sc = SeldonClient(
+        gateway=gateway_type,
+        transport=transport,
+        deployment_name=deployment_name,
+        payload_type=payload_type,
+        gateway_endpoint=host,
+        namespace=namespace,
+        client_return_type="dict",
+    )
+
+    Thread(
+        target=_start_input_file_worker, args=(q_in, input_data_path), daemon=True
+    ).start()
+
+    for _ in range(workers):
+        Thread(
+            target=_start_request_worker,
+            args=(q_in, q_out, data_type, sc, retries, batch_id),
+            daemon=True,
+        ).start()
+
+    Thread(
+        target=_start_output_file_worker, args=(q_out, output_data_path), daemon=True
+    ).start()
+
+    q_in.join()
+    q_out.join()
+
+    if benchmark:
+        print(f"Elapsed time: {time.time() - start_time}")
+
+
+def _start_input_file_worker(q_in: Queue, input_data_path: str) -> None:
+    """
+    Runs logic for the input file worker which reads the input file from filestore
+    and puts all of the lines into the input queue so it can be processed.
+
+    Parameters
+    ---
+    q_in
+        The queue to put all the data into for further processing
+    input_data_path
+        The local file to read the data from to be processed
+    """
+    input_data_file = open(input_data_path, "r")
+    enum_idx = 0
```

Comment:
Could use `enumerate` instead.

Reply:
Good idea - at the begininng I had to do it this way as I was making it a choice to provide either batch isntance IDs or enumerated IDs but now that I have both enumerate makes more sense

Reply:
It may be best to have it explicit for now as we may want to extend or remove the indices that are provided, I'm not really sure how useful it really would be adding the enum order, ideally users would be able to provide their own identifiers through request metadata